internal/resolution/variable_sources/entity/bundle_entity.go

```diff
@@ -0,0 +1,217 @@
+package entity
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"github.com/blang/semver/v4"
+	"github.com/operator-framework/deppy/pkg/deppy/input"
+	"github.com/operator-framework/operator-registry/alpha/property"
+)
+
+const PropertyBundlePath = "olm.bundle.path"
+
+type ChannelProperties struct {
+	property.Channel
+	Replaces  string   `json:"replaces,omitempty"`
+	Skips     []string `json:"skips,omitempty"`
+	SkipRange string   `json:"skipRange,omitempty"`
+}
+
+type PackageRequired struct {
+	property.PackageRequired
+	SemverRange *semver.Range `json:"-"`
+}
+
+type GVK property.GVK
+
+func (g GVK) String() string {
+	return fmt.Sprintf(`group:"%s" version:"%s" kind:"%s"`, g.Group, g.Version, g.Kind)
+}
+
+type GVKRequired property.GVKRequired
+
+func (g GVKRequired) String() string {
+	return fmt.Sprintf(`group:"%s" version:"%s" kind:"%s"`, g.Group, g.Version, g.Kind)
+}
+
+type BundleEntity struct {
+	*input.Entity
+
+	// these properties are lazy loaded as they are requested
+	bundlePackage     *property.Package
+	providedGVKs      []GVK
+	requiredGVKs      []GVKRequired
+	requiredPackages  []PackageRequired
+	channelProperties *ChannelProperties
+	semVersion        *semver.Version
+	bundlePath        string
+	mu                sync.RWMutex
+}
+
+func NewBundleEntity(entity *input.Entity) *BundleEntity {
+	return &BundleEntity{
+		Entity: entity,
+		mu:     sync.RWMutex{},
+	}
+}
+
+func (b *BundleEntity) PackageName() (string, error) {
+	if err := b.loadPackage(); err != nil {
+		return "", err
+	}
+	return b.bundlePackage.PackageName, nil
+}
+
+func (b *BundleEntity) Version() (*semver.Version, error) {
+	if err := b.loadPackage(); err != nil {
+		return nil, err
+	}
+	return b.semVersion, nil
+}
+
+func (b *BundleEntity) ProvidedGVKs() ([]GVK, error) {
+	if err := b.loadProvidedGVKs(); err != nil {
+		return nil, err
+	}
+	return b.providedGVKs, nil
+}
+
+func (b *BundleEntity) RequiredGVKs() ([]GVKRequired, error) {
+	if err := b.loadRequiredGVKs(); err != nil {
+		return nil, err
+	}
+	return b.requiredGVKs, nil
+}
+
+func (b *BundleEntity) RequiredPackages() ([]PackageRequired, error) {
+	if err := b.loadRequiredPackages(); err != nil {
+		return nil, err
+	}
+	return b.requiredPackages, nil
+}
+
+func (b *BundleEntity) ChannelName() (string, error) {
+	if err := b.loadChannelProperties(); err != nil {
+		return "", err
+	}
+	return b.channelProperties.ChannelName, nil
+}
+
+func (b *BundleEntity) ChannelProperties() (*ChannelProperties, error) {
+	if err := b.loadChannelProperties(); err != nil {
+		return nil, err
+	}
+	return b.channelProperties, nil
+}
+
+func (b *BundleEntity) BundlePath() (string, error) {
+	if err := b.loadBundlePath(); err != nil {
+		return "", err
+	}
+	return b.bundlePath, nil
+}
+
+func (b *BundleEntity) loadPackage() error {
+	if b.bundlePackage == nil {
```

Comment:
Is there no mutex lock required here unlike the other `loadX` funcs?

Reply:
Thank you!! Fixed